templates/cc/file.go

```diff
@@ -1,53 +1,65 @@
 package tpl
 
-const fileTpl = `// Code generated by protoc-gen-validate
+const moduleFileTpl = `// Code generated by protoc-gen-validate
 // source: {{ .InputPath }}
 // DO NOT EDIT!!!
 
-#pragma once
+#include "{{ .File.OutputPath.SetExt ".validate.h" }}"
 
-#include <string>
-#include <sstream>
+using std::string;
 
-#include "validate/validate.h"
-#include "{{ .File.OutputPath.SetExt ".h" }}"
+namespace pgv {
+namespace validate {
+{{ range .AllMessages }}
+{{- if not (disabled .) -}}
+bool CheckMessage(const {{ class . }}& m, std::string* err) {
+	return {{ package . }}::Validate(m, err);
+}
+{{ end }}
+{{ end }}
+
+{{ range (weakCheckMsgs .AllMessages) }}
+bool __attribute__((weak)) CheckMessage(const {{ . }}& m, std::string* err) {
```

Comment:
How portable is this between compilers?

Reply:
Clang supports weak linking. Searching around, it seems that there are some posts about certain early versions of Xcode having an issue, I think we will learn more when we try and integrate with Envoy. The other alternative we have I think is using static registration of checkers for different message types and doing dynamic dispatch; we can switch to that as a next step if this won't work nicely.

Reply:
I just validated that this works fine on OS X + clang. We know from the Envoy presubmits that it work with {gcc, clang} on Linux as well. I feel reasonably comfortable going forward with this approach as static weak linking as done by PGV here is fairly well understood an doesn't poke at any of the dark corners of weak link.